includes/class-validate-by-domain-options.php

```diff
@@ -0,0 +1,159 @@
+
+		add_options_page( 'Validate by Domain', 'Validate by Domain', 'manage_options', 'validate_by_domain', [
+			$this,
+			'plugin_options_page'
+		] );
+
+	}
+
+	/**
+	 * Register the plugin settings, create fields
+	 */
+	function plugin_settings_init() {
+
+		$args = array(
+			'sanitize_callback' => [ $this, 'sanitize_input' ],
+		);
+
+		register_setting( 'validate_by_domain', 'validate_by_domain_settings', $args );
+
+		add_settings_section(
+			'options_section',
+			__( '', 'WordPress' ),
+			'',
+			'validate_by_domain'
+		);
+
+		add_settings_field(
+			'validate_enable',
+			__( '', 'WordPress' ),
+			[ $this, 'enable_render' ],
+			'validate_by_domain',
+			'options_section'
+		);
+
+		add_settings_field(
+			'validate_role',
+			__( '', 'WordPress' ),
+			[ $this, 'role_render' ],
+			'validate_by_domain',
+			'options_section'
+		);
+
+		add_settings_field(
+			'validate_whitelist',
+			__( 'Whitelist', 'WordPress' ),
+			[ $this, 'whitelist_render' ],
+			'validate_by_domain',
+			'options_section'
+		);
+
+	}
+
+// todo: Check for the existence of the key before using it in a condition, in the render functions below (avoid Illegal string offset)
+
+	/**
+	 * Render the options page enable field
+	 */
+	function enable_render() {
+
+		$options = get_option( 'validate_by_domain_settings' );
+		?><!-- <input
+        type='checkbox'
+        name='validate_by_domain_settings[validate_enable]' <?php checked( $options['validate_enable'], 1 ); ?>
+        value='1'>--><?php
+	}
+
+	/**
+	 * Render the options page role selection field
+	 */
+	function role_render() {
+
+		$options = get_option( 'validate_by_domain_settings' );
+		?>
+        <!-- <select name='validate_by_domain_settings[validate_role]'>
+            <option value='1' <?php selected( $options['validate_role'], 1 ); ?>>Contributor</option>
+            <option value='2' <?php selected( $options['validate_role'], 2 ); ?>>Editor</option>
+            <option value='3' <?php selected( $options['validate_role'], 3 ); ?>>Author</option>
+            <option value='4' <?php selected( $options['validate_role'], 4 ); ?>>Subscriber</option>
+        </select> -->
+
+		<?php
+
+	}
+
+	/**
+	 * Render the options page whitelist field
+	 */
+	function whitelist_render() {
+
+		$options = get_option( 'validate_by_domain_settings' );
+		?>
+        <textarea cols='60' rows='15'
+                  name='validate_by_domain_settings[validate_whitelist]'><?php echo $options['validate_whitelist']; ?></textarea>
+		<?php
+
+	}
+
+	/**
+	 * @param $input
+	 * Sanitize the whitelist
+	 *
+	 * @return mixed|void
+	 */
+	function sanitize_input( $input ) {
+
+		// Create our array for storing the sanitized options
+		$output = [];
+
+		// Check if the current option has a value. If so, process it.
+		if ( isset( $input['validate_whitelist'] ) ) {
+
+			// Strip all HTML and PHP tags
+			$output['validate_whitelist'] = strip_tags( stripslashes( $input['validate_whitelist'] ) );
```

Comment:
this filters input, but do we not also need to validate input? If we don't then any string will suffice. Do we also want to look at unique values with `array_unique()` — May want to look at `filter_input`, `filter_input_array` and `FILTER_VALIDATE_REGEXP` 
![image](https://user-images.githubusercontent.com/2048170/37122899-14b9e27c-2217-11e8-9b2f-3b8b2805c409.png)

Reply:
Started down that route, but due to our sprints time constraints, I prefer to defer this to it's own issue. Along with what you've identified there's also other considerations for ensuring it's valid domain. Such as removing characters domains shouldn't have via FILTER_VALIDATE_URL or a regex since this would fail for urls with non-ascii chars. And then actually checking it's valid domain via checkdnsrr or gethostbyname )